pkg/controller/cert_control.go

```diff
@@ -0,0 +1,323 @@
+}
+
+func (rcc *realCertControl) sendCSR(ns string, instance string, rawCSR []byte, csrName string) (*capi.CertificateSigningRequest, error) {
+	var csr *capi.CertificateSigningRequest
+
+	// check for exist CSR, overwirte if it was created by operator, otherwise block the process
+	csr, err := rcc.getCSR(ns, instance, csrName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CSR for [%s/%s]: %s, error: %v", ns, instance, csrName, err)
+	}
+
+	if csr != nil {
+		glog.Infof("found exist CSR %s/%s created by tidb-operator, overwriting", ns, csrName)
+		delOpts := &types.DeleteOptions{TypeMeta: types.TypeMeta{Kind: "CertificateSigningRequest"}}
+		err := rcc.kubeCli.Certificates().CertificateSigningRequests().Delete(csrName, delOpts)
+		if err != nil {
+			return nil, fmt.Errorf("failed to delete exist old CSR for [%s/%s]: %s, error: %v", ns, instance, csrName, err)
+		}
+		glog.Infof("exist old CSR deleted for [%s/%s]: %s", ns, instance, csrName)
+		return rcc.sendCSR(ns, instance, rawCSR, csrName)
+	}
+
+	csr = &capi.CertificateSigningRequest{
+		TypeMeta: types.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: types.ObjectMeta{
+			Name:   csrName,
+			Labels: make(map[string]string),
+		},
+		Spec: capi.CertificateSigningRequestSpec{
+			Request: pem.EncodeToMemory(&pem.Block{
+				Type:    "CERTIFICATE REQUEST",
+				Headers: nil,
+				Bytes:   rawCSR,
+			}),
+			Usages: []capi.KeyUsage{
+				capi.UsageClientAuth,
+				capi.UsageServerAuth,
+			},
+		},
+	}
+
+	labelTemp := label.New()
+	csr.Labels[label.NamespaceLabelKey] = ns
+	csr.Labels[label.ManagedByLabelKey] = labelTemp[label.ManagedByLabelKey]
+	csr.Labels[label.InstanceLabelKey] = instance
+
+	resp, err := rcc.kubeCli.CertificatesV1beta1().CertificateSigningRequests().Create(csr)
+	if err != nil {
+		return resp, fmt.Errorf("failed to create CSR for [%s/%s]: %s, error: %v", ns, instance, csrName, err)
+	}
+	glog.Infof("CSR created for [%s/%s]: %s", ns, instance, csrName)
+	return resp, nil
+}
+
+func (rcc *realCertControl) approveCSR(csr *capi.CertificateSigningRequest) error {
+	csr.Status.Conditions = append(csr.Status.Conditions, capi.CertificateSigningRequestCondition{
+		Type:    capi.CertificateApproved,
+		Reason:  "AutoApproved",
+		Message: "Auto approved by TiDB Operator",
+	})
+	_, err := rcc.kubeCli.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+	if err != nil {
+		return fmt.Errorf("error updating approval for csr: %v", err)
+	}
+	return nil
+}
+
+/*
+func (rcc *realCertControl) RevokeCert() error {
+	return nil
+}
+*/
+/*
+func (rcc *realCertControl) RenewCert() error {
+	return nil
+}
+*/
+
+// LoadFromSecret loads cert and key from Secret matching the name
+func (rcc *realCertControl) LoadFromSecret(ns string, secretName string) ([]byte, []byte, error) {
+	secret, err := rcc.kubeCli.CoreV1().Secrets(ns).Get(secretName, types.GetOptions{})
```

Comment:
We should try to avoid access to apiserver directly, use the `SecretLister` to access the local cache instead. Refer https://github.com/pingcap/tidb-operator/blob/master/pkg/backup/util/util.go#L79